File: swap-nodes/swap-nodes.js
//Swap nodes that are next to each other. Ex.
//[x1 , y1 , x2 , y3] -> [y1 , x1 , y2, x2]
//This solution has a time compexity of O(n) where n
//is the length of the linked list.


/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if(head == null){
        return head;
    }
    if(head.next == null){
        return head;
    }  
    let itr1 = head;
    let itr2 = head?.next;
    while(itr2 != null){
        let temp = itr1.val;
        itr1.val = itr2.val;
        itr2.val = temp;
        itr1 = itr2?.next;
        itr2 = itr1?.next;
    }
    return head;
};

File: swap-nodes/swap-nodes.dart
//Swap every other node in a list then return the head.
//I did this the easy way by putting them into an array swapping
//them and setting the next val to the next.
//The time complexity if O(n) where n is the lenght of the list.
//Time: 258ms Beats: 77.78% 
//Memory: 144.5MB Beats: 22.22%
class Solution {
  ListNode? swapPairs(ListNode? head) {
      if(head == null || head?.next == null){
          return head;
      }
      List<ListNode?> nodes = [];
      while(head != null){
          nodes.add(head);
          head = head.next;
      }
      int itr = 0;
      while(itr < nodes.length - 1){
          ListNode? temp = nodes[itr];
          nodes[itr] = nodes[itr + 1];
          nodes[itr + 1] = temp;
          itr += 2;
      }
      for(int i = 0 ; i < nodes.length - 1; ++i){
          nodes[i]?.next = nodes[i + 1];
      }
      nodes[nodes.length - 1]?.next = null;
      return nodes[0];
  }
}
